gnome-initial-setup/pages/summary/gis-summary-page.c

```diff
@@ -236,6 +239,117 @@ done_cb (GtkButton *button, GisSummaryPage *page)
     }
 }
 
+static void
+initial_contact_app_signal (GDBusProxy *proxy,
+                            gchar *sender,
+                            gchar *signal_name,
+                            GVariant *parameters,
+                            gpointer user_data)
+{
+  GisSummaryPage *page = user_data;
+  GisSummaryPagePrivate *priv = gis_summary_page_get_instance_private (page);
+  const gchar *object, *property;
+  g_autoptr(GVariant) value;
+
+  if (g_strcmp0 (signal_name, "ObjectNotify") != 0)
+    return;
+
+  g_variant_get (parameters, "(&s&sv)", &object, &property, &value);
+
+  if (g_strcmp0 (object, "view.JSContext.globalParameters") != 0)
+    return;
+
+  if (g_strcmp0 (property, "unlocked") != 0)
+    return;
+
+  g_return_if_fail (g_variant_is_of_type (value, G_VARIANT_TYPE_BOOLEAN));
+
+  if (g_variant_get_boolean (value))
+    {
+      g_signal_handler_disconnect (priv->clippy_proxy,
+                                   priv->clippy_proxy_signal_id);
+      finish_fbe (page);
+    }
+}
+
+static void
+initial_contact_connect_to_clippy (GisSummaryPage *page)
+{
+  GisSummaryPagePrivate *priv = gis_summary_page_get_instance_private (page);
+  g_autoptr(GError) error = NULL;
+
+  g_dbus_proxy_call_sync (priv->clippy_proxy,
+                          "Connect",
+                          g_variant_new ("(sss)",
+                                         "view.JSContext.globalParameters",
+                                         "notify",
+                                         "unlocked"),
+                          G_DBUS_CALL_FLAGS_NONE,
+                          G_MAXINT,
+                          NULL,
+                          &error);
+
+  if (error != NULL)
+    {
+      g_critical ("Could not call Connect on DBus proxy for HackUnlock App: %s", error->message);
+
+      finish_fbe (page);
+      return;
+    }
+
+  priv->clippy_proxy_signal_id =
+    g_signal_connect (priv->clippy_proxy, "g-signal",
+                      G_CALLBACK (initial_contact_app_signal), page);
+}
+
+static gboolean
+initial_contact_connect_to_clippy_timeout (gpointer data)
+{
+  GisSummaryPage *page = data;
+  initial_contact_connect_to_clippy (page);
+  return G_SOURCE_REMOVE;
```

Comment:
I wonder if we should just rely on the 1 second timeout for this. Maybe a more robust strategy would be to keep trying for at least 3 times (3 secs). But eventually we'll fix the delay issue, so it's fine either way.

Reply:
Stupid question that I can't believe I don't know the answer to: if the callback returns `G_SOURCE_REMOVE`, is the ID returned by `g_timeout_add` still valid?

Reply:
It isn't! I need to reset it here. Good catch!